### ISSUE TITLE
Assertion errors running python -m unittest

### ISSUE THREAD

**Issue by mantier** [bamsumit/slayerPytorch#50]
Hi,

I am trying to set up slayer on an Azure VM. I set up a plain virtual environment with current packages from pip (torch==1.7.1, matplotlib==3.3.3, pyjaml==5.4.1, numpy==1.19.5, pickleshare==0.7.5, torchvision==0.8.2). I was able to compile SLAYER and I can run the examples.

However, many of the unittests fail. Any idea what the reason is?

One reason, that I think is unlikely, might be the fact that I do not have an X server, so matplotlib can not output anything.

OS: Ubuntu 18.04
Cuda: 10.1

Output:
```
(SLAYER_venv) azureuser@SLAYER-GPU:~/src/slayerPytorch/test$ python -m unittest
Max PSP kernel: 195.0                                                                                                   
Scaling neuron[scaleRho] by Max PSP
Kernel @slayerLoihi                                                                 
Max PSP kernel: 195.0                                                                                               
Scaling neuron[scaleRho] by Max PSP Kernel @slayerLoihi                                                                 
..F.FF......FF..F..F...FFFFFF....                                                                                      
======================================================================                                                  FAIL: testOut (test_slayer002.testSpike)                                                                                ----------------------------------------------------------------------                                                  Traceback (most recent call last):                                                                                        File "/home/azureuser/src/slayerPytorch/test/test_slayer002.py", line 196, in testOut                                     self.assertEqual(torch.norm(spikeGT - spikeOut), 0, 'Error mus be zero.')                                           AssertionError: tensor(157.0358, device='cuda:0', grad_fn=<CopyBackwards>) != 0 : Error mus be zero.                                                                                                                                            ======================================================================                                                  
FAIL: test (test_slayer003.testGradients)                                                                               
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer003.py", line 100, in test                                        self.assertTrue(errorLayer2 < 1e-4, 'Error must be close to zero.')                                                 
AssertionError: False is not true : Error must be close to zero.                                                                                                                                                                                
======================================================================                                                  
FAIL: test (test_slayer004.testSpikeClassifier)                                                                         
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer004.py", line 72, in test                                         self.assertEqual(predict.getClass(spikeOut), 21, 'Correct class is 21.')                                            AssertionError: tensor([24]) != 21 : Correct class is 21.                                                                                                                                                                                       
======================================================================                                                  
FAIL: testShift (test_slayer009.TestDelay)                                                                              
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer009.py", line 81, in testShift                                    self.assertEqual(netError, 0, 'Shift error must be zero.')                                                          
AssertionError: 38560289.8894043 != 0 : Shift error must be zero.                                                                                                                                                                               
======================================================================                                                  
FAIL: testShiftPerNeuron (test_slayer009.TestDelay)                                                                     
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer009.py", line 63, in testShiftPerNeuron                           self.assertEqual(netError, 0, 'Shift (per neuron) error must be zero.')                                             
AssertionError: 38543278.59399414 != 0 : Shift (per neuron) error must be zero.                                                                                                                                                                 
======================================================================                                                  
FAIL: test (test_slayer011.TestPspLayer)                                                                                
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer011.py", line 43, in test                                         
self.assertTrue(error < 1e-6, 'Result from Psp function and Psp layer must match.')                                 
AssertionError: tensor(False, device='cuda:0') is not true : Result from Psp function and Psp layer must match.                                                                                                                                 
======================================================================                                                  
FAIL: testValues (test_slayer012.TestPspFilter)                                                                         
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayer012.py", line 53, in testValues                                   
self.assertTrue(error < 1e-6, 'Result from Psp function and Psp filter bank must match.')                           
AssertionError: tensor(False, device='cuda:0') is not true : Result from Psp function and Psp filter bank must match.                                                                                                                           
======================================================================                                                  
FAIL: testOutput (test_slayerLoihi001.testLoihiSpikes)                                                                  
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 123, in testOutput                             self.assertTrue(error<1e-3, 'Output spike and ground truth must match.')                                            
AssertionError: False is not true : Output spike and ground truth must match.                                                                                                                                                                   
======================================================================                                                  
FAIL: testHidden (test_slayerLoihi001.testLoihiVoltage)                                                                 
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 129, in testHidden                             self.assertTrue(error<1e-3, 'Hidden voltage and ground truth must match.')                                          
AssertionError: False is not true : Hidden voltage and ground truth must match.                                                                                                                                                                 
======================================================================                                                  
FAIL: testOutput (test_slayerLoihi001.testLoihiVoltage)                                                                 
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                       
 File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 133, in testOutput                             self.assertTrue(error<1e-3, 'Hidden voltage and ground truth must match.')                                          
AssertionError: False is not true : Hidden voltage and ground truth must match.                                                                                                                                                                 
======================================================================                                                  
FAIL: testCurrent (test_slayerLoihi002.testRefDelay)                                                                    
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 126, in testCurrent                            self.assertTrue(error<1e-3, 'Current: cuda ouptut and ground truth must match.')                                    
AssertionError: False is not true : Current: cuda ouptut and ground truth must match.                                                                                                                                                           
======================================================================                                                  
FAIL: testSpikes (test_slayerLoihi002.testRefDelay)                                                                     
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 116, in testSpikes                             self.assertTrue(error<1e-3, 'Spikes: cuda ouptut and ground truth must match.')                                     
AssertionError: False is not true : Spikes: cuda ouptut and ground truth must match.                                                                                                                                                            
======================================================================                                                  
FAIL: testVoltage (test_slayerLoihi002.testRefDelay)                                                                    
----------------------------------------------------------------------                                                  
Traceback (most recent call last):                                                                                        
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 121, in testVoltage                            self.assertTrue(error<1e-3, 'Voltage: cuda ouptut and ground truth must match.')                                    
AssertionError: False is not true : Voltage: cuda ouptut and ground truth must match.                                                                                                                                                           
----------------------------------------------------------------------                                                  
Ran 33 tests in 119.874s                                                                                                                                                                                                                        
FAILED (failures=13)              
```

**Comment by bamsumit** [comment]
Hi,

My guess is your nvcc compiler and cuda runtime version does not match. It's not an issue with matplotlib.

What are the output of the following commands
`nvcc --version`
`nvidia-smi | grep CUDA`

**Comment by mantier** [comment]
Hi,

Unfortionally, the failing of tests persists after updating CUDA.

You were right, there was a mismatch of versions of the compiler (10.1) and runtime (11.1). I reinstalled CUDA using the official way. Now I get
```
(SLAYER_venv) azureuser@SLAYER-GPU:~/src/slayerPytorch/test$  nvcc --version                                      
nvcc: NVIDIA (R) Cuda compiler driver                                                                             
Copyright (c) 2005-2020 NVIDIA Corporation                                                                        
Built on Mon_Nov_30_19:08:53_PST_2020                                                                             
Cuda compilation tools, release 11.2, V11.2.67                                                                    
Build cuda_11.2.r11.2/compiler.29373293_0       
```
and
```
(SLAYER_venv) azureuser@SLAYER-GPU:~/src/slayerPytorch/test$  nvidia-smi | grep CUDA                              
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |      
```
I also added the export to the path variables to bash.rc:
```
export PATH=/usr/local/cuda-11.2/bin${PATH:+:${PATH}}
export LD_LIBRARY_PATH=/usr/local/cuda-11.2/${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
```

I recompiled SLAYER. First, I removed all files from the build (hopefully)
```
rm -r build dist
rm -r slayerCuda.egg-info/ slayerLoihiCuda.egg-info/ slayerSNN.egg-info/
pip uninstall slayerCuda slayerLoihiCuda slayerSNN
```
Then I recompiled with
```
python setup.py build
python setup.py install
```
The output of the unittests differs only marginally from the previous tests:
```
(SLAYER_venv) azureuser@SLAYER-GPU:~/src/slayerPytorch/test$ python -m unittest                                   
Max PSP kernel: 195.0                                                                                             
Scaling neuron[scaleRho] by Max PSP Kernel @slayerLoihi                                                           
Max PSP kernel: 195.0                                                                                             
Scaling neuron[scaleRho] by Max PSP Kernel @slayerLoihi                                                           
..F.FF......FF..F..F...FFFFFF....                                                                                 
======================================================================     
FAIL: testOut (test_slayer002.testSpike)                                                                          
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer002.py", line 196, in testOut                               
self.assertEqual(torch.norm(spikeGT - spikeOut), 0, 'Error mus be zero.')                                     
AssertionError: tensor(154.0745, device='cuda:0', grad_fn=<CopyBackwards>) != 0 : Error mus be zero.       
                                                                                                                         ======================================================================                                            
FAIL: test (test_slayer003.testGradients)                                                                         
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer003.py", line 100, in test                                  
self.assertTrue(errorLayer2 < 1e-4, 'Error must be close to zero.')                                           
AssertionError: False is not true : Error must be close to zero.              
                                                                                                                                                      ======================================================================                                            
FAIL: test (test_slayer004.testSpikeClassifier)                                                                   
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer004.py", line 72, in test                                   
self.assertEqual(predict.getClass(spikeOut), 21, 'Correct class is 21.')                                      
AssertionError: tensor([16]) != 21 : Correct class is 21.                                                                                                                                                                           

======================================================================                                            
FAIL: testShift (test_slayer009.TestDelay)                                                                        
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer009.py", line 81, in testShift                              
self.assertEqual(netError, 0, 'Shift error must be zero.')                                                    
AssertionError: 39332663.267456055 != 0 : Shift error must be zero.               
                                                                                                                                                  
======================================================================                                            
FAIL: testShiftPerNeuron (test_slayer009.TestDelay)                                                               
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                 
File "/home/azureuser/src/slayerPytorch/test/test_slayer009.py", line 63, in testShiftPerNeuron                     
self.assertEqual(netError, 0, 'Shift (per neuron) error must be zero.')                                       
AssertionError: 39316432.18884277 != 0 : Shift (per neuron) error must be zero.                                                                                                                                                     

======================================================================                                            
FAIL: test (test_slayer011.TestPspLayer)                                                                          
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer011.py", line 43, in test                                   
self.assertTrue(error < 1e-6, 'Result from Psp function and Psp layer must match.')                           
AssertionError: tensor(False, device='cuda:0') is not true : Result from Psp function and Psp layer must match.                                                                                                                     

======================================================================                                            
FAIL: testValues (test_slayer012.TestPspFilter)                                                                   -
---------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayer012.py", line 53, in testValues                             
self.assertTrue(error < 1e-6, 'Result from Psp function and Psp filter bank must match.')                     
AssertionError: tensor(False, device='cuda:0') is not true : Result from Psp function and Psp filter bank must match.                                                                                                                                                                                                                                 

======================================================================                                            
FAIL: testOutput (test_slayerLoihi001.testLoihiSpikes)                                                            
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 123, in testOutput                       
self.assertTrue(error<1e-3, 'Output spike and ground truth must match.')                                      
AssertionError: False is not true : Output spike and ground truth must match.                                                                                                                                                       

======================================================================                                            
FAIL: testHidden (test_slayerLoihi001.testLoihiVoltage)                                                           
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 129, in testHidden                       
self.assertTrue(error<1e-3, 'Hidden voltage and ground truth must match.')                                    
AssertionError: False is not true : Hidden voltage and ground truth must match.          
                                                                                                                                           
======================================================================                                            
FAIL: testOutput (test_slayerLoihi001.testLoihiVoltage)                                                           
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi001.py", line 133, in testOutput                       
self.assertTrue(error<1e-3, 'Hidden voltage and ground truth must match.')                                    
AssertionError: False is not true : Hidden voltage and ground truth must match.                                                                                                                                                     

======================================================================                                            
FAIL: testCurrent (test_slayerLoihi002.testRefDelay)                                                              
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 126, in testCurrent                      
self.assertTrue(error<1e-3, 'Current: cuda ouptut and ground truth must match.')                              
AssertionError: False is not true : Current: cuda ouptut and ground truth must match.                                                                                                                                               

======================================================================                                            
FAIL: testSpikes (test_slayerLoihi002.testRefDelay)                                                               
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 116, in testSpikes                       
self.assertTrue(error<1e-3, 'Spikes: cuda ouptut and ground truth must match.')                               
AssertionError: False is not true : Spikes: cuda ouptut and ground truth must match.                                                                                                                                                

======================================================================                                            
FAIL: testVoltage (test_slayerLoihi002.testRefDelay)                                                              
----------------------------------------------------------------------                                            
Traceback (most recent call last):                                                                                  
File "/home/azureuser/src/slayerPytorch/test/test_slayerLoihi002.py", line 121, in testVoltage                      
self.assertTrue(error<1e-3, 'Voltage: cuda ouptut and ground truth must match.')                              
AssertionError: False is not true : Voltage: cuda ouptut and ground truth must match.                                                                                                                                               

----------------------------------------------------------------------                                            
Ran 33 tests in 20.658s                                                                                                                                                                                                             
FAILED (failures=13)                                                                                              
```

**Comment by bamsumit** [comment]
There still seems to be mismatch between the cuda compiled binaries and runtime drivers. All of the errors are associated with cuda compiled codes. As far as I am aware, PyTorch is officially available for cuda 11.0. One way to verify the compatibility of nvcc and cuda runtime environment is compiling and running NVIDIA cuda samples: https://docs.nvidia.com/cuda/cuda-samples/index.html

**Alternatively**, an easier way could be to use docker container. Here is a minimal Dockerfile you can use:
```
from pytorch/pytorch:1.7.0-cuda11.0-cudnn8-devel

ENV DEBIAN_FRONTEND noninteractive

RUN apt-get update && apt-get install -y \
    python3-pip \
    gcc \
    vim

RUN cd /usr/local/bin && \
    ln -s /usr/bin/python3 python

RUN python3 -m pip install --upgrade pip
RUN pip3 install numpy pyyaml matplotlib h5py

ADD slayerPytorch /usr/local/slayerPytorch
RUN cd /usr/local/slayerPytorch &&\
    python setup.py install

ENV DEBIAN_FRONTEND teletype
```
**Note:** make sure slayerPytorch is in the docker context i.e. in the same folder as Dockerfile

**Comment by mantier** [comment]
First of all, thank you for your help. I finally figured it out.

The solution was to revert to a clean, new Azure VM (Standard NC6s_v2) with an Azure Data Science Virtual Machine (DSVM) for Linux (Ubuntu 18.04 LTS) and to use the provided configuration. It was simply:
```bash
conda activate py37_pytorch
git clone <SLAYER>
cd slayerPytorch
python setup.py install
cd test
python -m unittest
```
I get:
```
(py37_pytorch) azureuser@Slayer-Gpu-2:~/slayerPytorch/test$ python -m unittest                                                                                
Max PSP kernel: 195.0                                                                                                                                         
Scaling neuron[scaleRho] by Max PSP Kernel @slayerLoihi                                                                                                       
Max PSP kernel: 195.0                                                                                                                                         
Scaling neuron[scaleRho] by Max PSP Kernel @slayerLoihi                                                                                                       
.................................                                                                                                                             
----------------------------------------------------------------------                                                                                        
Ran 33 tests in 14.385s                                                                                                                                                                                                                                                                                                     
OK             
```

Here's my lessons learned; maybe it will help someone.
- I have CUDA drivers 11.2 (from `nvidia-smi`) and nvcc 10.1 (from `nvcc --version`), and it does not seem to be a problem.
- From inside python, `torch.version.cuda` gives me 10.2 as an output, which also does not seem to be a problem.
- I have the newest torch installed that is available from pip, which is 1.7.1, which seems to be not compatible with nvcc 11.x
- Try to stick with the standard configuration as long as possible
- If you have changed a lot in the configuration, try a clean machine

I am not completely sure where I made a mistake in the first place. I assume the problem was that I tried to work with a standard pyhton virtual environment and not with anaconda. I don't like anaconda too much, I always get in trouble with it. However, in this case it made sense to keep it.

Looking forward to work with SLAYER now!

**Comment by bamsumit** [comment]
Great. Thanks for sharing how you fixed it.

**Comment by kkkkkkip** [comment]
> There still seems to be mismatch between the cuda compiled binaries and runtime drivers. All of the errors are associated with cuda compiled codes. As far as I am aware, PyTorch is officially available for cuda 11.0. One way to verify the compatibility of nvcc and cuda runtime environment is compiling and running NVIDIA cuda samples: https://docs.nvidia.com/cuda/cuda-samples/index.html
> 
> **Alternatively**, an easier way could be to use docker container. Here is a minimal Dockerfile you can use:
> 
> ```
> from pytorch/pytorch:1.7.0-cuda11.0-cudnn8-devel
> 
> ENV DEBIAN_FRONTEND noninteractive
> 
> RUN apt-get update && apt-get install -y \
>     python3-pip \
>     gcc \
>     vim
> 
> RUN cd /usr/local/bin && \
>     ln -s /usr/bin/python3 python
> 
> RUN python3 -m pip install --upgrade pip
> RUN pip3 install numpy pyyaml matplotlib h5py
> 
> ADD slayerPytorch /usr/local/slayerPytorch
> RUN cd /usr/local/slayerPytorch &&\
>     python setup.py install
> 
> ENV DEBIAN_FRONTEND teletype
> ```
> 
> **Note:** make sure slayerPytorch is in the docker context i.e. in the same folder as Dockerfile

ADD slayerPytorch /usr/local/slayerPytorch       Do I need to set the path to store slayerpytoch myself
